src/herder/HerderSCPDriver.cpp

```diff
@@ -1285,4 +1286,79 @@ HerderSCPDriver::TxSetValidityKeyHash::operator()(
+        APPLICATION_SPECIFIC_NOMINATION_LEADER_ELECTION_PROTOCOL_VERSION);
+    if (unsupportedProtocol || !cfg.VALIDATOR_WEIGHT_CONFIG.has_value() ||
+        cfg.FORCE_OLD_STYLE_LEADER_ELECTION)
+    {
+        // Fall back on old weight algorithm if any of the following are true:
+        // 1. The network has not yet upgraded to
+        //    APPLICATION_SPECIFIC_NOMINATION_LEADER_ELECTION_PROTOCOL_VERSION,
+        // 2. The node is using manual quorum set configuration, or
+        // 3. The node has the FORCE_OLD_STYLE_LEADER_ELECTION flag
+        //    set
+        return SCPDriver::getNodeWeight(nodeID, qset, isLocalNode);
+    }
+
+    ValidatorWeightConfig const& vwc =
+        mApp.getConfig().VALIDATOR_WEIGHT_CONFIG.value();
+
+    auto entryIt = vwc.mValidatorEntries.find(nodeID);
+    if (entryIt == vwc.mValidatorEntries.end())
+    {
+        // This shouldn't be possible as the validator entries should contain
+        // all validators in the config. For this to happen, `getNodeWeight`
+        // would have to be called with a non-validator `nodeID`. Throw if
+        // building tests, and otherwise fall back on the old algorithm.
+        throw std::runtime_error(
+            fmt::format(FMT_STRING("Validator entry not found for node {}"),
+                        toShortString(nodeID)));
+    }
+
+    ValidatorEntry const& entry = entryIt->second;
+    if (entry.mQuality == ValidatorQuality::VALIDATOR_LOW_QUALITY)
+    {
+        // Special case: low quality validators have weight `0`
+        return 0;
+    }
+
+    auto homeDomainSizeIt = vwc.mHomeDomainSizes.find(entry.mHomeDomain);
+    if (homeDomainSizeIt == vwc.mHomeDomainSizes.end())
+    {
+        // This shouldn't be possible as the home domain sizes should contain
+        // all home domains in the config. For this to happen, `getNodeWeight`
+        // would have to be called with a non-validator, or the config parser
+        // would have to allow a validator without a home domain. Throw if
+        // building tests, and otherwise fall back on the old algorithm.
+        throw std::runtime_error(
+            fmt::format(FMT_STRING("Home domain size not found for domain {}"),
+                        entry.mHomeDomain));
+    }
+
+    // First, divide the theoretical maximum organization weight by the total
+    // number of nodes in that organization
+    uint64 const numerator = UINT64_MAX / homeDomainSizeIt->second;
+
+    // Then, compute the number of quality levels between the highest quality
+    // org and this node's org
+    int const qualityDelta = static_cast<int>(vwc.mHighestQuality) -
+                             static_cast<int>(entry.mQuality);
+    releaseAssert(qualityDelta >= 0);
+
+    // Finally, divide the theoretical max weight for nodes in this org
+    // (`numerator`) by `10^{qualityDelta}` to get the final weight.
+    uint64 const denominator =
+        static_cast<uint64>(std::pow(10.0, qualityDelta));
```

Comment:
The new weight function looks good to me, but going back to the design doc, it's actually different from the one proposed, isn't it? Is it intended?

Reply:
Oh wow, this is embarrassing. Dropping the `|O_q|` term from the design doc was not intentional.

When we added the multiplicative factor in the denominator, the importance of the `|O_q|` term was significantly reduced, to the point that I completely forgot about it.  As you noted, the weight function without it still satisfies the fairness criteria we care about.

However, dropping the term overweights `MEDIUM` quality validators a little bit beyond what we originally planned, so I felt it was worth fixing. Adding the term back involved calculating quality-level weights (`w_q` in the design doc) in `Config::setValidatorWeightConfig` and tweaking `HerderSCPDriver::getNodeWeight` to use those new weights. Please take another look at those two functions now that this is fixed.